locator/abstract_replication_strategy.cc

```diff
@@ -234,7 +234,20 @@ abstract_replication_strategy::get_address_ranges(const token_metadata& tm) cons
 future<std::unordered_multimap<inet_address, dht::token_range>>
 abstract_replication_strategy::get_address_ranges(const token_metadata& tm, inet_address endpoint) const {
     std::unordered_multimap<inet_address, dht::token_range> ret;
+    if (!tm.is_member(endpoint)) {
+        co_return ret;
+    }
     for (auto& t : tm.sorted_tokens()) {
+        // This is a fast path for everywhere_topology to avoid calculating
+        // natural endpoints, since any node that is part of the the ring
+        // will be responsible for all tokens.
+        if (get_type() == replication_strategy_type::everywhere_topology) {
```

Comment:
Maybe we should follow the approach taken by other functions i.e. make `get_address_ranges` a virtual function and override it in everywhere_strategy?

Reply:
I tired this. However, unlike get_natural_endpoints which is a virtual and has a very different implementation for different strategies. The implementation of get_address_ranges is almost identical, making it virtual and override in everywhere_strategy is just duplicating the code and makes it harder to understand how get_address_ranges works.

Reply:
Why duplicate? We continue the loop here `if (get_type() == replication_strategy_type::everywhere_topology)`
so the rest of the inner loop won't be there.

Reply:
Yes, but we still have the check for is_member and the loop over tm.sorted_tokens() in both versions. And we need to reason about the motivation to override for everywhere_topology and the difference of the implementations in two different places. 

Again, I am aiming to backport to 4.6. Minimum change is better IMO.

Reply:
okay

Reply:
Could you extract the `if (get_type() == ...)` outside the loop? We don't need to check this again for every token, we can check once.

Reply:
yup. done